backends/s3/s3.go

```diff
@@ -217,6 +222,15 @@ func (b *Backend) Delete(ctx context.Context, name string) error {
 	if err = handleErrorResponse(err); err != nil {
 		metricCallErrors.WithLabelValues("delete").Inc()
 	}
+	if b.opt.UseUpdateMarker {
+		if err := b.doStore(ctx, UpdateMarkerFilename, []byte(name)); err != nil {
```

Comment:
This introduces buggy behaviour.

Previously the update marker would contain the filename of any newly added file, signaling that something changed. These file are assumed to be unique between instances (no two instances will try to write the same name). If this assumption is not correct, the update marker optimisation MUST NOT be used.

But now with this change you are also setting this to indicate that you deleted the file. Consider the following sequence of events:

- Instance A writes foo.txt and updates the marker to 'foo.txt'
- Instance D lists the files and records the last marker as 'foo.txt'
- Instance B writes bar.txt and updates the marker to 'bar.txt'
- Instance C removes 'foo.txt' and updates the marker to 'foo.txt'
- Instance D lists the files again and determines that nothing has changed, because the marker did not change.

Oops.

If you want to update the marker in this case, you have to set it to a different value.

Perhaps we can set the marker to a combination of:

- the blob name affected
- the nanosecond timestamp of the change (or a hash of the contents)

This would allow the mechanism to also work correctly for both updates and deletes. For delete you could also add an indicator that the change was a deletion.

A hash would provide slightly stronger guarantees than a timestamp, but it adds overhead (especially for bigger blobs) and we already have the forced periodic reload to handle these.

Reply:
Thank you!
I'll probably use a combination of the blob name, the operation type (store/delete) and the nanosecond timestamp. Then comparing the ETag should be enough, wouldn't it?
Also, as using the hash of the content would be safer, maybe replacing the timestamp with the ETag of the blob that was just uploaded makes sense?

Reply:
An ETag would be fine too, but I don't know if we can rely on every S3 API implementation returning one. Perhaps use it when available, and fall back to a timestamp if not?